Added user settings dialog to customize snippet folder location and creating "Samples" files on start up

Havagan opened the pull request:
Added user settings dialog to customize snippet folder location and creating "Samples" files on start up

Issue #36

aelij commented:
Thanks for the contribution! However, I would prefer not to have unnecessary dialogs. Changing the location of the documents can be done from a button at the bottom of the left document pane.

Also, I'm planning on changing the samples so they won't get extracted to disk at all, so that option is also redundant. For now, you can simply not extract after changing the location.

Havagan commented on src/RoslynPad/UserSettingsDialog.xaml.cs:
```diff
@@ -0,0 +1,157 @@
+        private void DocumentPathBrowse_Click(object sender, RoutedEventArgs e)
+        {
+            // TODO: Review use of System.Windows.Forms reference.
+            var dialog = new System.Windows.Forms.FolderBrowserDialog();
```
The OpenFileDialog in WPF requires you to select a file and won't let you select a folder reference. The two workaround I see posted more frequently use one of two methods:

* System.Windows.Forms.FolderBrowserDialog from the .NET framework --> https://msdn.microsoft.com/en-us/library/system.windows.forms.folderbrowserdialog.aspx
* CommonOpenFileDialog from the Windows API Code pack --> https://github.com/aybe/Windows-API-Code-Pack-1.1

Did you have a preference or know of a pure WPF way to implement the select a folder dialog?

I'm not a WPF developer (one reason I wanted to contribute to this project was to learn some WPF) -- sorry if this seems simplistic.

Havagan commented:
I implemented the changes you requested. Icon is in place at bottom left. Document pane rebinds when folder changed. (Updated Documents, etc. to notify when changed and updated the Document tree view to update when the property changes.)

The only pending question is the one noted above about proper way to do a Folder browse/select in WPF vs a File browse/select 

![selectdocumentfolder](https://cloud.githubusercontent.com/assets/2288850/20636262/a5ad4008-b337-11e6-8253-ea8e1de6980c.png)
dialog.

aelij commented on src/RoslynPad/DocumentViewModel.cs:
```diff
@@ -19,16 +19,76 @@ internal class DocumentViewModel : NotificationObject
+            // Currently rebinding documents pane.
+            // Dialog requesting restart? Auto-restart app? Rebind documents pane? What happens to open documents?
+
+            var dialog = new System.Windows.Forms.FolderBrowserDialog()
```
We'll use the `Avalon.Windows.Dialogs.FolderBrowserDialog`, which doesn't require win forms reference (it's a library of mine already referenced in RP)

aelij commented on src/RoslynPad/DocumentViewModel.cs:
```diff
@@ -19,16 +19,76 @@ internal class DocumentViewModel : NotificationObject
+        }
+
+        public async void EditUserDocumentPath()
+        {
```
This method should be in the main view model

aelij commented on src/RoslynPad/MainViewModel.cs:
```diff
@@ -67,7 +67,16 @@ public MainViewModel()
 
             DocumentRoot = CreateDocumentRoot();
+
+            DocumentRoot.PropertyChanged += (sender, args) =>
```
Not required if `EditUserDocumentPath` is move here

aelij commented on src/RoslynPad/DocumentViewModel.cs:
```diff
@@ -19,16 +19,76 @@ internal class DocumentViewModel : NotificationObject
+            var result = dialog.ShowDialog();
+            if (result == System.Windows.Forms.DialogResult.OK)
+            {
+                await MainViewModel.CloseAllDocuments();
```
We definitely don't want to close documents just because we changed folders. The current documents will continue to be saved in the original location, and that's fine.

aelij commented on src/RoslynPad/MainWindow.xaml:
```diff
@@ -144,11 +151,28 @@
                                    HorizontalAlignment="Center" />
                     </StackPanel>
                 </Button>
+
```
This should be inside the docking manager where the tree view is. Wrap it in a dock panel and dock this stack panel to the bottom.

aelij commented on src/RoslynPad/MainWindow.xaml:
```diff
@@ -144,11 +151,28 @@
+                                                VerticalAlignment="Bottom"
+                                                Background="Transparent"
+                                                Style="{StaticResource {x:Static ToolBar.ButtonStyleKey}}"
+                                                Click="EditDocumentPathButton_Click">
```
Use a command instead of the click event

aelij commented on src/RoslynPad/DocumentViewModel.cs:
```diff
@@ -19,16 +19,76 @@ internal class DocumentViewModel : NotificationObject
+                SaveDocumentPath(dialog.SelectedPath);
+            }
+
+            //var dialog = new Microsoft.Win32.OpenFileDialog()
```
Remove unused code

aelij commented on src/RoslynPad/MainViewModel.cs:
```diff
@@ -233,9 +254,16 @@ public void CreateNewDocument()
+
+        public async Task CloseAllDocuments()
+        {
+            // can't modify the collection while enumerating it.
```
You can use `OpenDocuments.ToArray()`. Also remove the comment - this is very common knowledge in .NET :)

aelij commented on src/RoslynPad/MainWindow.xaml:
```diff
@@ -144,11 +151,28 @@
+                                            HorizontalAlignment="Left"
+                                            VerticalAlignment="Bottom"
+                                            Margin="5">
+                    <Button Name="EditDocumentPathButton" HorizontalAlignment="Left"
```
Remove the button's Name

aelij commented on src/RoslynPad/MainWindow.xaml:
```diff
@@ -122,8 +123,14 @@
+                                                           ItemContainerStyle="{StaticResource DocumentItemContainerStyle}">
+                                        <controls:TreeListView.ContextMenu>
+                                            <ContextMenu>
+                                                <MenuItem Name="DocumentsContextMenu_OpenFolder" Header="_Open Folder in Windows Explorer" Click="DocumentsContextMenu_OpenFolder_Click"></MenuItem>
```
Remove the menu item's Name

aelij replied:
Use a command instead of the Click event

aelij commented on src/RoslynPad/RoslynPad.csproj:
```diff
@@ -65,6 +65,7 @@
     <Reference Include="System" />
     <Reference Include="System.ComponentModel.Composition" />
     <Reference Include="System.Data" />
+    <Reference Include="System.Windows.Forms" />
```
Remove this reference

aelij commented on src/RoslynPad/DocumentViewModel.cs:
```diff
@@ -19,16 +19,76 @@ internal class DocumentViewModel : NotificationObject
         }
 
-        public string Path { get; }
+        public string Path { get; internal set; }
```
Remove this setter. We'll simply create a new `DocumentRoot` when changing folders.

aelij commented on src/RoslynPad/DocumentViewModel.cs:
```diff
@@ -19,16 +19,76 @@ internal class DocumentViewModel : NotificationObject
         public MainViewModel MainViewModel { get; }
         public bool IsFolder { get; }
 
+        internal static string GetUserDocumentPath()
```
Move to `MainViewModel` and pass the path parameter to the constructor

aelij commented:
Thanks for all your work! I'll take it from here :)
